src/librustc_typeck/check/regionck.rs

```diff
@@ -1449,15 +1449,15 @@ impl<'a, 'gcx, 'tcx> RegionCtxt<'a, 'gcx, 'tcx> {
         // the problem is to add `T: 'r`, which isn't true. So, if there are no
         // inference variables, we use a verify constraint instead of adding
         // edges, which winds up enforcing the same condition.
-        let needs_infer = projection_ty.trait_ref.needs_infer();
+        let needs_infer = projection_ty.trait_ref(self.tcx).needs_infer();
         if env_bounds.is_empty() && needs_infer {
             debug!("projection_must_outlive: no declared bounds");
 
-            for component_ty in projection_ty.trait_ref.substs.types() {
+            for component_ty in projection_ty.trait_ref(self.tcx).substs.types() {
```

Comment:
Did this substitution globally (I wasn't sure about the semantic difference between the two substs, but someone will have to do a full review anyway with ATC when they become different).